### PR TITLE
commit module5

### PR DESCRIPTION
**Pull request by domrk98git**
please help review code and need help with production cmd

### CODE REVIEW

**Review by Jonesey13 — CHANGES_REQUESTED** [review]
I don't think there's anything relevant to module 5 here? Branches 4, 5 & 6 are all at the same commit.

For this module I'm expecting to see a `dockerfile` at the very least.

File: Vagrantfile [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  config.trigger.after :up do |trigger|
+    trigger.name = "Launching App"
+    trigger.info = "Running the TODO app setup script"
+    trigger.run_remote = {privileged: false, inline: "
+      cd /vagrant
+      poetry install
+      poetry run flask run --host 0.0.0.0
+    "}
+  end
+end
```
Comment (Jonesey13) [inline Vagrantfile:35]: This file is out of scope for Module 5

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+RUN curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python
+
+
+# copy across app code
+WORKDIR ${APP_INSTALL}
+COPY . /app
+RUN poetry install
+
+# list all copied files
+RUN ls -la ./
```
Comment (Jonesey13) [inline Dockerfile:21]: This line doesn't actually do anything

**Review by Jonesey13 — CHANGES_REQUESTED** [review]
This is the right idea but you're missing a few key steps:

- There is no .dockerignore file
- There are no instructions on how to build and run the dockerfile
- The dockerfile is not a multi-stage build with development and production targets.

File: .dockerignore [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+.DS_Store
+
+### Virtual environment ###
+env/
+.venv/
+
+### Configuration files ###
+.env
+### module4
+.vagrant
```
Comment (hugh-emerson) [inline .dockerignore:11]: This is the right idea but more things could be added. E.g `Vagrantfile`, `.git`, `.gitignore`, `__pycache__` `Dockerfile`

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+
+# define entry point , and default launch command
+
+# for development
+FROM base as development
+EXPOSE 5002
```
Comment (hugh-emerson) [inline Dockerfile:28]: This is wrong. This command means that you expect the **container** to have something running on port 5002, but you don't. `flask run` will start a server listening to port 5000 by default.

The image doesn't specify what port you use on your machine (the host machine) because it can't know. That's determined entirely by what you put in the `docker run` command.

Similarly the production image's expose is also wrong. It should be `EXPOSE 8000` because that's what gunicorn will run on (you could specify a different port with e.g. `--bind 0.0.0.0:8080`)

File: README.md [hunk 5]
```diff
@@ -50,3 +50,60 @@ You should see output similar to the following:
  * Debugger PIN: 226-556-590
 ```
 Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser to view the app.
+
+## Add development build stage to dockerfile
+ docker build --target development --tag module5:dev .
+
+ ## Running the App in dev environment
+
+(base) RaviDommatas-Mac-mini:DevOps-Course-Starter rkdommata$ docker run -p 5000:5000 --env-file .env module5
```
Comment (hugh-emerson) [inline README.md:59]: Its a bit weird to include your full prompt `RaviDommatas-Mac-mini:DevOps-Course-Starter rkdommata`

File: README.md [hunk 6]
```diff
@@ -50,3 +50,60 @@ You should see output similar to the following:
...
+ * Running on http://0.0.0.0:5000/ (Press CTRL+C to quit)
+ * Restarting with stat
+ * Debugger is active!
+ * Debugger PIN: 307-202-561
+
+with mount bind on docker run :
+docker run -p 5000:5000 --mount type=bind,source="$(pwd)",target=/app --env-file .env module5
+ ## Running the App in PROD environment
+ ## added below entry to pyproject.toml
+ gunicorn = "20.0.4"
```
Comment (hugh-emerson) [inline README.md:72]: Not sure what these two lines are doing here.

Also you can check the formatting with the preview button in the top right of VSCode or you can see it in GitHub https://github.com/domrk98git/DevOps-Course-Starter/blob/7b4d00ee61a84a78099bfa1f3820e7bc0da18f48/README.md

File: README.md [hunk 7]
```diff
@@ -50,3 +50,60 @@ You should see output similar to the following:
...
+Accessed production at http://0.0.0.0:5002
+
+## mount bind on docker run : prod and dev
+(base) RaviDommatas-Mac-mini:DevOps-Course-Starter rkdommata$ docker run -p 5003:8000 --mount type=bind,source="$(pwd)",target=/todo_app.app --env-file .env  module5:prod
+[2021-02-20 04:13:01 +0000] [7] [INFO] Starting gunicorn 20.0.4
+[2021-02-20 04:13:01 +0000] [7] [INFO] Listening at: http://0.0.0.0:8000 (7)
+[2021-02-20 04:13:01 +0000] [7] [INFO] Using worker: sync
+[2021-02-20 04:13:01 +0000] [11] [INFO] Booting worker with pid: 11
+
+(base) RaviDommatas-Mac-mini:DevOps-Course-Starter rkdommata$ docker run -p 5000:5000 --mount type=bind,source="$(pwd)",target=/todo_app.app --env-file .env  module5
```
Comment (hugh-emerson) [inline README.md:97]: You already had this instruction higher up.

**Review by hugh-emerson — APPROVED** [review]
I've left some suggestions but generally looks good. You can merge when you're ready but I haven't reviewed the included changes relevant to modules 4/6. You should just merge the module 5 changes.
